### PR TITLE
Add support for virtual network interfaces

### PR DESCRIPTION
**Pull request by jukkar**
With these commits, it is possible to stack network interfaces on top of each other. This can be used to implement IP tunneling and VPN support. There is a IPIP driver implementation included, it supports ipip, ipip6, ip6ip and ip6ip6 tunneling.

### CODE REVIEW

File: tests/net/virtual/Kconfig [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Private config options for echo-server sample app
+
+# Copyright (c) 2018 Intel Corporation
```
Comment (rlubos) [inline tests/net/virtual/Kconfig:3]: 2021?

File: tests/net/virtual/Kconfig [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "Networking echo-server sample application"
```
Comment (rlubos) [inline tests/net/virtual/Kconfig:6]: This description doesn't seem right.

File: subsys/net/l2/virtual/ipip/ipip.c [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+
+	if (ctx->status) {
+		return -EALREADY;
+	}
+
+	ctx->status = true;
+
+	NET_DBG("Starting iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
```
Comment (rlubos) [inline subsys/net/l2/virtual/ipip/ipip.c:96]: Should this comment be here? I'd expect it more in a sample code than an actual tunneling driver.

File: subsys/net/l2/virtual/ipip/ipip.c [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+
+	if (!ctx->status) {
+		return -EALREADY;
+	}
+
+	ctx->status = false;
+
+	NET_DBG("Stopping iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
```
Comment (rlubos) [inline subsys/net/l2/virtual/ipip/ipip.c:115]: Same here.

**Review by rlubos — COMMENTED** [review]
Pretty neat

File: subsys/net/l2/virtual/ipip/ipip.c (hunk 3 above)
Comment (jukkar) [reply subsys/net/l2/virtual/ipip/ipip.c:96]: > Should this comment be here? I'd expect it more in a sample code than an actual tunneling driver.

Yeah, a left over from some other ideas. I will remove this.

File: include/net/virtual.h [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+		 * notices carrier ON/OFF situation. We must not create another
+		 * network management event from inside management handler thus
+		 * we use worker thread to trigger the UP/DOWN event.
+		 */
+		struct k_work work;
+
+		/** Network interface that is detecting carrier ON/OFF event.
+		 */
+		struct net_if *iface;
+	} carrier_mgmt;
```
Comment (jukkar) [inline include/net/virtual.h:151]: I think I will remove the carrier handling as that is handled by the real network interface.

File: include/net/virtual_mgmt.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_OFF,
+};
+
+#define NET_EVENT_VIRTUAL_INTERFACE_CARRIER_ON				\
+	(_NET_VIRTUAL_INTERFACE_EVENT |					\
+	 NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_ON)
+
+#define NET_EVENT_VIRTUAL_INTERFACE_CARRIER_OFF				\
+	(_NET_VIRTUAL_INTERFACE_EVENT |					\
+	 NET_EVENT_VIRTUAL_INTERFACE_CMD_CARRIER_OFF)
```
Comment (jukkar) [inline include/net/virtual_mgmt.h:51]: These would go away too.

File: include/net/virtual_mgmt.h [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+ */
+#if defined(CONFIG_NET_L2_VIRTUAL_MGMT)
+void virtual_interface_mgmt_raise_carrier_off_event(struct net_if *iface);
+#else
+static inline
+void virtual_mgmt_interface_raise_carrier_off_event(struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+}
+#endif
```
Comment (jukkar) [inline include/net/virtual_mgmt.h:125]: And these are not really needed.

File: subsys/net/l2/virtual/virtual.c [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+	struct virtual_interface_context *ctx = net_if_l2_data(iface);
+
+	if (k_work_pending(&ctx->carrier_mgmt.work)) {
+		return -EAGAIN;
+	}
+
+	handle_carrier(ctx, iface, carrier_off);
+
+	return 0;
+}
```
Comment (jukkar) [inline subsys/net/l2/virtual/virtual.c:180]: Would go away

File: subsys/net/l2/virtual/virtual_mgmt.c [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+void virtual_interface_mgmt_raise_carrier_on_event(struct net_if *iface)
+{
+	net_mgmt_event_notify(NET_EVENT_VIRTUAL_INTERFACE_CARRIER_ON, iface);
+}
+
+void virtual_interface_mgmt_raise_carrier_off_event(struct net_if *iface)
+{
+	net_mgmt_event_notify(NET_EVENT_VIRTUAL_INTERFACE_CARRIER_OFF, iface);
+}
```
Comment (jukkar) [inline subsys/net/l2/virtual/virtual_mgmt.c:24]: Would be removed

**Review by rlubos — APPROVED** [review]
LGTM, awesome work 👍

**Comment by jukkar** [comment]
Are there anyone else able to review?

File: subsys/net/ip/net_shell.c [hunk 10]
```diff
@@ -284,6 +284,53 @@ static void print_supported_ethernet_capabilities(
...
+				"IPv4,");
+	}
+
+	if (net_if_flag_is_set(iface, NET_IF_IPV6)) {
+		pos += snprintk(str + pos, sizeof(str) - pos,
+				"IPv6,");
+	}
+
+	/* get rid of last ',' character */
+	str[pos - 1] = '\0';
```
Comment (mniestroj) [inline subsys/net/ip/net_shell.c:329]: What if there are no flags? Is that possible?

**Review by mniestroj — APPROVED** [review]
I didn't get deep into virtual interface implementation, all the bits around seem fine. Just one comment, which is a possible out of buffer access.

File: tests/net/virtual/testcase.yaml [hunk 11]
```diff
@@ -0,0 +1,25 @@
+common:
+  depends_on: netif
+  min_ram: 32
+  tags: net tunnel virtual
```
Comment (nashif) [inline tests/net/virtual/testcase.yaml:4]: please add integration_platforms and list the platforms where this should run on in CI for pull requests.

**Review by nashif — CHANGES_REQUESTED** [review]

File: subsys/net/ip/net_shell.c (hunk 10 above)
Comment (jukkar) [reply subsys/net/ip/net_shell.c:329]: > What if there are no flags? Is that possible?

In theory that could be possible but in practice not. Anyway, I will add checks for the value.

**Comment by jukkar** [comment]
@nashif  integration_platforms added to tests, could you recheck

**Comment by jukkar** [comment]
ping @nashif

**Comment by jukkar** [comment]
ping @nashif

**Review by nashif — APPROVED** [review]

**Pull request merged**